WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift

```diff
@@ -333,9 +366,89 @@ private extension ProductDetailPreviewViewModel {
     }
 }
 
+// MARK: - Categories
+//
+private extension ProductDetailPreviewViewModel {
+    @MainActor
+    func synchronizeAllCategories() async throws {
+        try await withCheckedThrowingContinuation { continuation in
+            stores.dispatch(ProductCategoryAction.synchronizeProductCategories(siteID: siteID,
+                                                                               fromPageNumber: Default.firstPageNumber,
+                                                                               onCompletion: { result in
+                continuation.resume()
+            }))
+        }
+    }
+
+    @MainActor
+    func addCategories(_ names: [String]) async throws -> [ProductCategory] {
+        guard names.isNotEmpty else {
+            return []
+        }
+        return try await withCheckedThrowingContinuation { continuation in
+            stores.dispatch(ProductCategoryAction.addProductCategories(siteID: siteID,
+                                                                       names: names,
+                                                                       parentID: nil,
+                                                                       onCompletion: { result in
+                continuation.resume(with: result)
+            }))
+        }
+    }
+}
+
+// MARK: - Tags
+//
+private extension ProductDetailPreviewViewModel {
+    @MainActor
+    func synchronizeAllTags() async throws {
+        try await withCheckedThrowingContinuation { continuation in
+            stores.dispatch(ProductTagAction.synchronizeAllProductTags(siteID: siteID,
+                                                                       onCompletion: { result in
+                continuation.resume()
+            }))
+        }
+    }
+
+    @MainActor
+    func addTags(_ names: [String]) async throws -> [ProductTag] {
+        guard names.isNotEmpty else {
+            return []
+        }
+        return try await withCheckedThrowingContinuation { continuation in
+            stores.dispatch(ProductTagAction.addProductTags(siteID: siteID,
+                                                            tags: names,
+                                                            onCompletion: { result in
+                continuation.resume(with: result)
+            }))
+        }
+    }
+}
+
 // MARK: - Saving product
 //
 private extension ProductDetailPreviewViewModel {
+    @MainActor
+    func saveLocalCategoriesAndTags(_ product: Product) async throws -> Product {
```

Comment:
Nit: It may be worth adding a comment explaining this method too.

Reply:
Added a comment in https://github.com/woocommerce/woocommerce-ios/pull/10864/commits/85b3b3c599938256db03eb7282e60771e8eff2d2